### DISCUSSION TITLE
Setting a max time for a refresh Token chain?

### DISCUSSION THREAD

**Discussion by tsoestini** [ory/hydra#2886] [Q&A] [answered]
Is there a way to set a maximum amount of time a refreshToken chain can be used?

I see the TTL_REFRESH_TOKEN setting. However, if I set it to 30 days, it appears that a client can refresh its refresh tokens every 7 days for an entire year.

Is there a way to say that that whole chain expires after 6 months and the user needs to re-authorize the client?

**Comment by vinckr** [comment, marked as answer]
Hello @tsoestini 

AFAIK there is no way to achieve this at the moment. 
It has come up several times before however, here is quite a long discussion going over the feature and how it can be achieved: https://github.com/ory/hydra/issues/1529
Maybe that helps to put it into perspective or give some outlook.

**Reply by tsoestini** [reply]
Thanks for the confirmation and for the link to the other thread.

If I'm understanding correctly, it sounds like a lot of the concerns in the other thread are around having client-dependent lengths for refresh tokens.

Possibly what is involved with people talking past each other in that thread is three different conceptions of refresh tokens:

Group A: Understands that refresh tokens can currently be refreshed forever, and fully expects anyone who has a refresh token to do so (as far as security assumptions are concerned).

Group B: Understands that refresh tokens can currently be refreshed forever, but plans to use the refresh tokens in a single page app, with the understanding that when the user does not have that page open in their browser, then the refresh token might naturally expire.

Group C: Expects refresh tokens to not be refreshable forever, and expects that the chain will time out at some point.

Now, Group A rightly argues that it doesn't make much sense to worry about adjusting refresh token duration on a per-client basis because whatever the duration is, it effectively means forever. Whether you have a 30 day token you refresh every day forever or a 90 day token you refresh every day forever or a 4 hour token you refresh every hour forever, you still have forever access.

Group C mistakenly argues that it is important to adjust refresh token duration on a per-client basis, because in Group C's view there is a very real effect of different refresh token durations. For some clients, it is acceptable that they have access forever. But others they only want to have access for a month, others a week, and others only a day.

Group B adds a bit of confusion to the arguments. On the one hand, they agree with conclusion of Group C that refresh token durations should be configured on a per-client basis, but they disagree with the premise of Group C that refresh tokens can't in principle be refreshed forever. Meanwhile, they see a valid use in the refresh token for extending access during the user's continuous use of a web page, though Group A finds that use to be off-label from the intended use and would rather see that extension managed at the login or consent layer.

I guess I'm coming in here as a sort of Group D, which probably means I'm adding a *ton* of confusion to the arguments. I think I see things from the point of view of Group A as to how things *currently* are, but also see that the way Group C expects things already work is indicative of a missing feature.

I think of it sort of like a "statue of limitations". Yes, you can normally refresh forever, but let's figure out what the tolerance is for the system-level cleanup. For example, if we can say that no refresh token chain can go on for more than a year, that provides a number of benefits:
(a) any database entries in hydra older than a year can be safely deleted.
(b) any access logs that we keep that are older than a year can be safely deleted.
(c) we only have to trace back at most one year for any customer service queries we get from customers trying to track down access to their account.

Without a maximum lifetime for refresh chains, the only way to "clean out" the system is to delete all the secrets and generate a new one, but this has the side effect that even people who went through the consent flow a couple hours ago have to give their consent again.

Using secret rotation doesn't work to "clean out" the system, because a chain from 5 years ago will just get rotated to the new secrets at its next refresh.

Using /oauth/flush doesn't work to "clean out" the system, because it doesn't flush refresh tokens either.

**Reply by tsoestini** [reply]
I opened a issue to see if a 1-year maximum lifetime can be added: https://github.com/ory/hydra/issues/2889 .

**Reply by vinckr** [reply]
Thanks a lot for the write up and going to the effort of grouping all the different positions in that lengthy discussion 🙌  

It is probably hard to find a very convincing technical reason why it should be the one way or the other, its a more fuzzy question of operational security - which can vary a lot between the different usage modes. 
Your proposal to make the max lifetime configurable does sound reasonable to me!
